### PR TITLE
Improve action execution flow.

### PR DESCRIPTION
**Pull request by Thisara-Welmilla**
Task issue:
- https://github.com/wso2/product-is/issues/20931

Improve the action execution flow to execute actions that need to be apply by providing action id list.

### CODE REVIEW

File: components/action-mgt/org.wso2.carbon.identity.action.execution/src/main/java/org/wso2/carbon/identity/action/execution/ActionExecutorServiceImpl.java [hunk 1]
```diff
@@ -84,26 +84,74 @@ public boolean isExecutionEnabled(ActionType actionType) {
         return ActionExecutorConfig.getInstance().isExecutionForActionTypeEnabled(actionType);
     }
 
+    /**
+     * Resolve the actions that need to be executed for thr given action types and execute them.
```
Comment (malithie) [inline components/action-mgt/org.wso2.carbon.identity.action.execution/src/main/java/org/wso2/carbon/identity/action/execution/ActionExecutorServiceImpl.java:88]: typo -> tyr

File: components/action-mgt/org.wso2.carbon.identity.action.execution/src/main/java/org/wso2/carbon/identity/action/execution/ActionExecutorServiceImpl.java [hunk 2]
```diff
@@ -84,26 +84,74 @@ public boolean isExecutionEnabled(ActionType actionType) {
...
+
+        try {
+            // As of now only one action is allowed.
+            return execute(getActionByActionId(actionType, actionIds[0], tenantDomain), eventContext);
+        } catch (ActionExecutionRuntimeException e) {
+            // todo: add to diagnostics
+            LOG.debug("Skip executing actions for action type: " + actionType.name(), e);
+            return new ActionExecutionStatus(ActionExecutionStatus.Status.FAILED, eventContext);
+        }
+    }
```
Comment (malithie) [inline components/action-mgt/org.wso2.carbon.identity.action.execution/src/main/java/org/wso2/carbon/identity/action/execution/ActionExecutorServiceImpl.java:131]: cover unit tests for this method

Comment (Thisara-Welmilla) [reply components/action-mgt/org.wso2.carbon.identity.action.execution/src/main/java/org/wso2/carbon/identity/action/execution/ActionExecutorServiceImpl.java]: Units test are added now.

File: components/action-mgt/org.wso2.carbon.identity.action.execution/src/main/java/org/wso2/carbon/identity/action/execution/ActionExecutorServiceImpl.java (hunk 1 above)
Comment (Thisara-Welmilla) [reply components/action-mgt/org.wso2.carbon.identity.action.execution/src/main/java/org/wso2/carbon/identity/action/execution/ActionExecutorServiceImpl.java]: Addressed

File: components/action-mgt/org.wso2.carbon.identity.action.execution/src/main/java/org/wso2/carbon/identity/action/execution/ActionExecutorServiceImpl.java [hunk 3]
```diff
@@ -85,26 +85,77 @@ public boolean isExecutionEnabled(ActionType actionType) {
...
+     * @param eventContext  The event context of the corresponding flow.
+     * @param tenantDomain  Tenant domain.
+     * @return Action execution status.
+     */
+    public ActionExecutionStatus execute(ActionType actionType, String[] actionIds, Map<String, Object> eventContext,
+                                         String tenantDomain) throws ActionExecutionException {
 
+        try {
+            // As of now only one action is allowed.
+            if (actionIds.length != 1) {
```
Comment (malithie) [inline components/action-mgt/org.wso2.carbon.identity.action.execution/src/main/java/org/wso2/carbon/identity/action/execution/ActionExecutorServiceImpl.java:126]: Let's do the validation to make sure this array is not null and not empty

Comment (malithie) [reply components/action-mgt/org.wso2.carbon.identity.action.execution/src/main/java/org/wso2/carbon/identity/action/execution/ActionExecutorServiceImpl.java:126]: Basically extract to a validation method similar to https://github.com/wso2/carbon-identity-framework/pull/5902/files#diff-30dbfcba373317b490119e2dbd937558324dc1d02e3cd745febf0d9631b792ebR101

File: components/action-mgt/org.wso2.carbon.identity.action.execution/src/main/java/org/wso2/carbon/identity/action/execution/ActionExecutorServiceImpl.java [hunk 4]
```diff
@@ -85,26 +85,77 @@ public boolean isExecutionEnabled(ActionType actionType) {
...
+                throw new ActionExecutionException("Only support list with single action Id.");
+            }
+            return execute(getActionByActionId(actionType, actionIds[0], tenantDomain), eventContext);
+        } catch (ActionExecutionRuntimeException e) {
+            // todo: add to diagnostics
+            LOG.debug("Skip executing actions for action type: " + actionType.name(), e);
```
Comment (malithie) [inline components/action-mgt/org.wso2.carbon.identity.action.execution/src/main/java/org/wso2/carbon/identity/action/execution/ActionExecutorServiceImpl.java:132]: It's fine to log an error here as this is a real error case.
Let's follow the same pattern similar to https://github.com/wso2/carbon-identity-framework/blob/b7e7c8d2d8339f0b99e57ded2b7244160e18d365/components/action-mgt/org.wso2.carbon.identity.action.execution/src/main/java/org/wso2/carbon/identity/action/execution/ActionExecutorServiceImpl.java#L168-L170

File: components/action-mgt/org.wso2.carbon.identity.action.execution/src/main/java/org/wso2/carbon/identity/action/execution/ActionExecutorServiceImpl.java [hunk 5]
```diff
@@ -85,26 +85,77 @@ public boolean isExecutionEnabled(ActionType actionType) {
...
+    }
+
+    private Action getActionByActionId(ActionType actionType, String actionId, String tenantDomain)
+            throws ActionExecutionRuntimeException {
+
+        try {
+            return ActionExecutionServiceComponentHolder.getInstance().getActionManagementService().getActionByActionId(
+                    Action.ActionTypes.valueOf(actionType.name()).getActionType(), actionId, tenantDomain);
+        } catch (ActionMgtException e) {
+            throw new ActionExecutionRuntimeException("Error occurred while retrieving action by action Id .", e);
```
Comment (malithie) [inline components/action-mgt/org.wso2.carbon.identity.action.execution/src/main/java/org/wso2/carbon/identity/action/execution/ActionExecutorServiceImpl.java:158]: we can cover up this scenario in tests cases. Let's do that

File: components/action-mgt/org.wso2.carbon.identity.action.execution/src/main/java/org/wso2/carbon/identity/action/execution/ActionExecutorServiceImpl.java (hunk 4 above)
Comment (Thisara-Welmilla) [reply components/action-mgt/org.wso2.carbon.identity.action.execution/src/main/java/org/wso2/carbon/identity/action/execution/ActionExecutorServiceImpl.java:130]: Improved this to be done with validateActionIdList() methods.

File: components/action-mgt/org.wso2.carbon.identity.action.execution/src/main/java/org/wso2/carbon/identity/action/execution/ActionExecutorServiceImpl.java (hunk 5 above)
Comment (Thisara-Welmilla) [reply components/action-mgt/org.wso2.carbon.identity.action.execution/src/main/java/org/wso2/carbon/identity/action/execution/ActionExecutorServiceImpl.java]: Unit tests are added now.

File: components/action-mgt/org.wso2.carbon.identity.action.execution/src/main/java/org/wso2/carbon/identity/action/execution/ActionExecutorServiceImpl.java (hunk 3 above)
Comment (Thisara-Welmilla) [reply components/action-mgt/org.wso2.carbon.identity.action.execution/src/main/java/org/wso2/carbon/identity/action/execution/ActionExecutorServiceImpl.java]: Improved this to be done with  new validateActionIdList() methods.

File: components/action-mgt/org.wso2.carbon.identity.action.execution/src/test/java/org/wso2/carbon/identity/action/execution/ActionExecutorServiceImplTest.java [hunk 6]
```diff
@@ -133,40 +133,79 @@ public void testActionExecuteFailureWhenMultipleActionsAvailableForActionType()
         actionExecutorService.execute(ActionType.PRE_ISSUE_ACCESS_TOKEN, any(), any());
     }
 
+    @Test(expectedExceptions = ActionExecutionException.class,
+            expectedExceptionsMessageRegExp = "No action Ids found for action type: PRE_ISSUE_ACCESS_TOKEN")
+    public void testActionExecuteWithActionIdsFailureWhenNotSingleActionIdAvailable() throws Exception {
```
Comment (malithie) [inline components/action-mgt/org.wso2.carbon.identity.action.execution/src/test/java/org/wso2/carbon/identity/action/execution/ActionExecutorServiceImplTest.java:138]: This method name is not quite clear.
Isn't it 'testActionExecuteWithActionIdsFailureWhenActionIdListIsEMPTY'

Comment (Thisara-Welmilla) [reply components/action-mgt/org.wso2.carbon.identity.action.execution/src/test/java/org/wso2/carbon/identity/action/execution/ActionExecutorServiceImplTest.java]: Addressed.

**Review by malithie — APPROVED** [review]

**Pull request merged**